// @gear-js/api/lib/cjs/index.js
'use strict';

require('@polkadot/api-augment');
var consts = require('./consts.js');
var Events = require('./events/Events.js');
var GearEventData = require('./events/GearEventData.js');
var GearApi = require('./GearApi.js');
var GearProgram = require('./GearProgram.js');
var Keyring = require('./Keyring.js');
var createType = require('./metadata/create-type.js');
var is = require('./metadata/is.js');
var metadata = require('./metadata/metadata.js');
var programMetadata = require('./metadata/programMetadata.js');
var stateMetadata = require('./metadata/stateMetadata.js');
var rpc = require('./types/interfaces/message/rpc.js');
require('@polkadot/types/lookup');
var address = require('./utils/address.js');
var createPayload = require('./utils/create-payload.js');
var generate = require('./utils/generate.js');
var getExtrinsic = require('./utils/getExtrinsic.js');
var json = require('./utils/json.js');
var prefixes = require('./utils/prefixes.js');
var regexp = require('./utils/regexp.js');
var replyCode = require('./utils/reply-code.js');
var signature = require('./utils/signature.js');
var string = require('./utils/string.js');
var types = require('./utils/types.js');
var validate = require('./utils/validate.js');
require('node:assert');
require('@polkadot/types');
var reply = require('./wasm/reply.js');



Object.defineProperty(exports, "SPEC_VERSION", {
	enumerable: true,
	get: function () { return consts.SPEC_VERSION; }
});
exports.VARA_SS58_FORMAT = consts.VARA_SS58_FORMAT;
exports.GearEvents = Events.GearEvents;
exports.GearEventData = GearEventData.GearEventData;
exports.GearApi = GearApi.GearApi;
exports.BaseGearProgram = GearProgram.BaseGearProgram;
exports.Program = GearProgram.Program;
exports.GearKeyring = Keyring.GearKeyring;
exports.CreateType = createType.CreateType;
exports.isProgramMeta = is.isProgramMeta;
exports.isStateMeta = is.isStateMeta;
exports.GearMetadata = metadata.GearMetadata;
Object.defineProperty(exports, "Lang", {
	enumerable: true,
	get: function () { return programMetadata.Lang; }
});
Object.defineProperty(exports, "MetadataVersion", {
	enumerable: true,
	get: function () { return programMetadata.MetadataVersion; }
});
exports.ProgramMetadata = programMetadata.ProgramMetadata;
exports.StateMetadata = stateMetadata.StateMetadata;
exports.getStateMetadata = stateMetadata.getStateMetadata;
exports.PayloadFilter = rpc.PayloadFilter;
exports.decodeAddress = address.decodeAddress;
exports.encodeAddress = address.encodeAddress;
exports.encodePayload = createPayload.encodePayload;
exports.getRegistry = createPayload.getRegistry;
exports.generateCodeHash = generate.generateCodeHash;
exports.generateProgramId = generate.generateProgramId;
exports.generateVoucherId = generate.generateVoucherId;
exports.getExtrinsic = getExtrinsic.getExtrinsic;
exports.isJSON = json.isJSON;
exports.toJSON = json.toJSON;
exports.getIdsFromKeys = prefixes.getIdsFromKeys;
exports.REGULAR_EXP = regexp.REGULAR_EXP;
Object.defineProperty(exports, "EReplyCode", {
	enumerable: true,
	get: function () { return replyCode.EReplyCode; }
});
Object.defineProperty(exports, "ESimpleExecutionError", {
	enumerable: true,
	get: function () { return replyCode.ESimpleExecutionError; }
});
Object.defineProperty(exports, "ESimpleUnavailableActorError", {
	enumerable: true,
	get: function () { return replyCode.ESimpleUnavailableActorError; }
});
Object.defineProperty(exports, "ESuccessReply", {
	enumerable: true,
	get: function () { return replyCode.ESuccessReply; }
});
exports.ErrorReplyReason = replyCode.ErrorReplyReason;
exports.ExecutionErrorReason = replyCode.ExecutionErrorReason;
exports.ReplyCode = replyCode.ReplyCode;
exports.SuccessReplyReason = replyCode.SuccessReplyReason;
exports.UnavailableActorErrorReason = replyCode.UnavailableActorErrorReason;
exports.signatureIsValid = signature.signatureIsValid;
exports.toCamelCase = string.toCamelCase;
exports.convertString = types.convertString;
exports.getTypeAndPayload = types.getTypeAndPayload;
exports.joinTypePath = types.joinTypePath;
exports.transformTypes = types.transformTypes;
exports.typeIsGeneric = types.typeIsGeneric;
exports.typeIsString = types.typeIsString;
exports.validateCodeId = validate.validateCodeId;
exports.validateGasLimit = validate.validateGasLimit;
exports.validateMailboxItem = validate.validateMailboxItem;
exports.validateProgramId = validate.validateProgramId;
exports.validateValue = validate.validateValue;
exports.getGrReply = reply.getGrReply;
