// sails-js/lib/cjs/index.cjs
var prefix = require('./prefix.cjs');
var transactionBuilder = require('./transaction-builder.cjs');
var queryBuilder = require('./query-builder.cjs');
var utils = require('./utils.cjs');
var consts = require('./consts.cjs');
var nonZero = require('./primitives/non-zero.cjs');
var hash = require('./primitives/hash.cjs');
var sails = require('./sails.cjs');



exports.getCtorNamePrefix = prefix.getCtorNamePrefix;
exports.getFnNamePrefix = prefix.getFnNamePrefix;
exports.getServiceNamePrefix = prefix.getServiceNamePrefix;
exports.TransactionBuilder = transactionBuilder.TransactionBuilder;
exports.QueryBuilder = queryBuilder.QueryBuilder;
exports.throwOnErrorReply = utils.throwOnErrorReply;
exports.ZERO_ADDRESS = consts.ZERO_ADDRESS;
exports.NonZeroU128 = nonZero.NonZeroU128;
exports.NonZeroU16 = nonZero.NonZeroU16;
exports.NonZeroU256 = nonZero.NonZeroU256;
exports.NonZeroU32 = nonZero.NonZeroU32;
exports.NonZeroU64 = nonZero.NonZeroU64;
exports.NonZeroU8 = nonZero.NonZeroU8;
exports.ActorId = hash.ActorId;
exports.CodeId = hash.CodeId;
exports.H160 = hash.H160;
exports.H256 = hash.H256;
exports.MessageId = hash.MessageId;
exports.Sails = sails.Sails;
